revocation/revocation.go

```diff
@@ -101,49 +110,142 @@ func NewWithOptions(opts Options) (Validator, error) {
+// NOTE: The certificate chain is expected to be in the order of leaf to root.
 func (r *revocation) Validate(certChain []*x509.Certificate, signingTime time.Time) ([]*result.CertRevocationResult, error) {
 	return r.ValidateContext(context.Background(), ValidateContextOptions{
 		CertChain:            certChain,
 		AuthenticSigningTime: signingTime,
 	})
 }
 
-// ValidateContext checks the revocation status for a certificate chain using
-// OCSP and returns an array of CertRevocationResults that contain the results
-// and any errors that are encountered during the process
+// ValidateContext checks the revocation status for a certificate chain using OCSP and
+// CRL if OCSP is not available. It returns an array of CertRevocationResults
+// that contain the results and any errors that are encountered during the
+// process.
+//
+// This function tries OCSP and falls back to CRL when:
+// - OCSP is not supported by the certificate
+// - OCSP returns an unknown status
 //
-// TODO: add CRL support
-// https://github.com/notaryproject/notation-core-go/issues/125
+// NOTE: The certificate chain is expected to be in the order of leaf to root.
 func (r *revocation) ValidateContext(ctx context.Context, validateContextOpts ValidateContextOptions) ([]*result.CertRevocationResult, error) {
 	if len(validateContextOpts.CertChain) == 0 {
 		return nil, result.InvalidChainError{Err: errors.New("chain does not contain any certificates")}
 	}
+	certChain := validateContextOpts.CertChain
 
-	return ocsp.CheckStatus(ocsp.Options{
-		CertChain:        validateContextOpts.CertChain,
-		CertChainPurpose: r.certChainPurpose,
-		SigningTime:      validateContextOpts.AuthenticSigningTime,
-		HTTPClient:       r.httpClient,
-	})
+	if err := x509util.ValidateChain(certChain, r.certChainPurpose); err != nil {
+		return nil, err
+	}
+
+	ocspOpts := ocsp.CertCheckStatusOptions{
+		HTTPClient:  r.ocspHTTPClient,
+		SigningTime: validateContextOpts.AuthenticSigningTime,
+	}
+
+	crlOpts := crl.CertCheckStatusOptions{
+		HTTPClient:  r.crlHTTPClient,
+		SigningTime: validateContextOpts.AuthenticSigningTime,
+	}
+
+	panicChan := make(chan any, len(certChain))
+	defer close(panicChan)
+
+	certResults := make([]*result.CertRevocationResult, len(certChain))
+	var wg sync.WaitGroup
+	for i, cert := range certChain[:len(certChain)-1] {
+		switch {
+		case ocsp.Supported(cert):
+			// do OCSP check for the certificate
+			wg.Add(1)
+
+			go func(i int, cert *x509.Certificate) {
+				defer wg.Done()
+				defer func() {
+					if r := recover(); r != nil {
+						panicChan <- r
+					}
+				}()
+
+				ocspResult := ocsp.CertCheckStatus(cert, certChain[i+1], ocspOpts)
+				if ocspResult != nil && ocspResult.Result == result.ResultUnknown && crl.Supported(cert) {
+					// try CRL check if OCSP result is unknown
+					crlResult := crl.CertCheckStatus(ctx, cert, certChain[i+1], crlOpts)
+
+					// insert OCSP result into CRL result
+					crlResult.ServerResults = ocspResult.ServerResults
```

Comment:
I think it had better to be another PR because we will refactor the OCSP package. 
Added an issue #222

Reply:
However, `crlResult.ServerResults = ocspResult.ServerResults` doesn't look correct to me. This can be easily interpreted as `assigning the server result of ocsp directly to the server result of crl`.

Renaming `crlResult` to `result` makes more sense:
```suggestion
					// insert OCSP result into revocation result
					result.ServerResults = ocspResult.ServerResults
```

Reply:
Updated.